helix-core/src/main/java/org/apache/helix/model/Message.java

```diff
@@ -966,4 +966,46 @@ public boolean isValid() {
     }
     return true;
   }
+
+  /**
+   * This class is for categorizing state transition messages based on certain properties, and generating
+   */
+  public static class MessageInfo {
+    public String _msgType;
+    public String _resourceName;
+    public String _fromState;
+    public String _toState;
+
+    public enum MessageIdentifierBase {
+      PER_RESOURCE, PER_STATE_TRANSITION_TYPE
+    }
+
+    public MessageInfo(Message message) {
+      _msgType = message.getMsgType();
+      _resourceName = message.getResourceName();
+      _fromState = message.getFromState();
+      _toState = message.getToState();
+    }
+
+    public String getMessageIdentifier(MessageIdentifierBase basis) {
+      String delimiter = ".";
+      if (_msgType == null) {
+        return null;
+      }
+      String identifier = _msgType;
+      if (basis.ordinal() >= MessageIdentifierBase.PER_RESOURCE.ordinal()) {
```

Comment:
How was the ordinal work? The number larger means the position higher? What if this non-existing type?

Reply:
The original code is to concatenate strings for each level resource -> state transition type... This is for the same purpose.
But as we discussed earlier, some levels can be parallel, not necessarily hierarchical. For example we can have resourceName_stateTransitionType, resourceName_rebalanceType, where stateTransitionType and rebalanceType are parallel. I modified the code of comparing the ordinal, please take another look. Thanks.